server/dns_server.ml

```diff
@@ -671,32 +679,34 @@ let update_data trie zone (prereq, update) =
      Error Rcode.YXRRSet) >>= fun () ->
   if Dns_trie.equal trie trie' then
     (* should this error out? - RFC 2136 3.4.2.7 says NoError at the end *)
-    Ok (trie, None)
-  else match Dns_trie.lookup zone Soa trie, Dns_trie.lookup zone Soa trie' with
-    | Ok oldsoa, Ok soa when Soa.newer ~old:oldsoa soa ->
-      Ok (trie', Some (zone, soa))
-    | _, Ok soa ->
-      let soa = { soa with Soa.serial = Int32.succ soa.Soa.serial } in
-      let trie'' = Dns_trie.insert zone Soa soa trie' in
-      Ok (trie'', Some (zone, soa))
-    | Ok oldsoa, Error _ ->
-      (* zone removal!? *)
-      let serial = Int32.succ oldsoa.Soa.serial in
-      Ok (trie', Some (zone, { oldsoa with Soa.serial }))
-    | Error o, Error n ->
-      Log.warn (fun m -> m "shouldn't happen: %a no soa in old %a and new %a"
-                   Domain_name.pp zone Dns_trie.pp_e o Dns_trie.pp_e n);
-      Ok (trie', None)
+    Ok (trie, [])
+  else
+    Domain_name.Set.fold (fun zone acc ->
+        acc >>= fun (trie', zones) ->
+        match Dns_trie.lookup zone Soa trie, Dns_trie.lookup zone Soa trie' with
+        | Ok oldsoa, Ok soa when Soa.newer ~old:oldsoa soa ->
+          Ok (trie', (zone, soa) :: zones)
+        | _, Ok soa ->
+          let soa = { soa with Soa.serial = Int32.succ soa.Soa.serial } in
+          let trie'' = Dns_trie.insert zone Soa soa trie' in
+          Ok (trie'', (zone, soa) :: zones)
+        | Ok oldsoa, Error _ ->
+          (* zone removal!? *)
```

Comment:
Is this worth a debug log here, as it seems an unusual case to encounter in the wild?

Reply:
depends on deployment etc. -- atm that's a bit untested (and not advertised) that you can dynamically add/remove zones. once that is tested (and used and deployed), I'm keen on getting rid of this comment, which imho serves its purpose for now.

Reply:
Sounds good